README


This archive contains the distribution MooX-LvalueAttribute,
version 0.15:

  Provides Lvalue accessors to Moo class attributes

This software is copyright (c) 2013 by Damien "dams" Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.15      2013-09-02 17:08:34 Europe/Paris

  * disable XS optimization per attribute using lvalue, not per entire class

0.14      2013-09-02 15:14:18 Europe/Paris

  * Fix tests when Class::XSAccessor is installed

0.13      2013-08-27 17:55:49 Europe/Paris

0.12      2013-08-21 16:43:14 Europe/Paris

  * can be used from a Moo::Role

0.11      2013-08-21 15:54:27 Europe/Paris

0.10      2013-02-11 23:01:08 Europe/Paris

  * first implementation


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Provides Lvalue accessors to Moo class attributes",
  "dist_author" => [
    "Damien \"dams\" Krotkine"
  ],
  "dist_name" => "MooX-LvalueAttribute",
  "dist_version" => "0.15",
  "license" => "perl",
  "module_name" => "MooX::LvalueAttribute",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Hash::Util::FieldHash::Compat" => 0,
    "Moo" => "1.000008",
    "Variable::Magic" => "0.52",
    "strictures" => 1
  },
  "script_files" => [],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Devel::Hide" => 0,
    "Test::More" => "0.88",
    "blib" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/MooX/LvalueAttribute.pm
#
# This file is part of MooX-LvalueAttribute
#
# This software is copyright (c) 2013 by Damien "dams" Krotkine.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package MooX::LvalueAttribute;
{
  $MooX::LvalueAttribute::VERSION = '0.15';
}
use strictures 1;

# ABSTRACT: Provides Lvalue accessors to Moo class attributes

require Moo;
require Moo::Role;

our %INJECTED_IN_ROLE;
our %INJECTED_IN_CLASS;

sub import {
    my $class = shift;
    my $target = caller;

    if ($Moo::Role::INFO{$target} && $Moo::Role::INFO{$target}{is_role}) {

        # We are loaded from a Moo role
        $Moo::Role::INFO{$target}{accessor_maker} ||= do {
            require Method::Generate::Accessor;
            Method::Generate::Accessor->new
          };
        Moo::Role->apply_roles_to_object(
            $Moo::Role::INFO{$target}{accessor_maker},
            'Method::Generate::Accessor::Role::LvalueAttribute',
        );
        $INJECTED_IN_ROLE{$target} = 1;

    } elsif ($Moo::MAKERS{$target} && $Moo::MAKERS{$target}{is_class}) {

        # We are loaded from a Moo class
        if ( !$INJECTED_IN_CLASS{$target} ) {
            Moo::Role->apply_roles_to_object(
              Moo->_accessor_maker_for($target),
              'Method::Generate::Accessor::Role::LvalueAttribute',
            );
            $INJECTED_IN_CLASS{$target} = 1;        
        }
    } else {
        die "MooX::LvalueAttribute can only be used in Moo classes or Moo roles.";        
    }

}


1;

__END__
=pod

=head1 NAME

MooX::LvalueAttribute - Provides Lvalue accessors to Moo class attributes

=head1 VERSION

version 0.15

=head1 SYNOPSIS

=head2 From a Moo class

  package App;
  use Moo;
  use MooX::LvalueAttribute;
  
  has name => (
    is => 'rw',
    lvalue => 1,
  );

  # Elsewhere

  my $app = App->new(name => 'foo');
  $app->name = 'Bar';
  print $app->name;  # Bar

=head2 From a Moo role

  package MyRole;
  use Moo::Role;
  use MooX::LvalueAttribute;

  has name => (
    is => 'rw',
    lvalue => 1,
  );

  package App;
  use Moo;
  with('MyRole');

  # Elsewhere

  my $app = App->new(name => 'foo');
  $app->name = 'Bar';
  print $app->name;  # Bar

=head1 DESCRIPTION

This modules provides Lvalue accessors to your Moo attributes. It won't break
Moo's encapsulation, and will properly call any accessor method modifiers,
triggers, builders and default values creation. It can be used from a Moo class
or role.

It means that instead of writing:

  $object->name("Foo");

you can use:

  $object->name = "Foo"; 

=head1 ATTRIBUTE SPECIFICATION

To enable Lvalue access to your attribute, simply use C<MooX::LvalueAttribute>
in the class or role, and add:

  lvalue => 1,

in the attribute specification (see synopsis).

=head1 NOTE ON IMPLEMENTATION

The implementation doesn't use AUTOLOAD, nor TIESCALAR. Instead, it uses a
custom accessor and C<Variable::Magic>, which is faster and cheaper than the
tie / AUTOLOAD mechanisms.

=head1 AUTHOR

Damien "dams" Krotkine

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Damien "dams" Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Provides Lvalue accessors to Moo class attributes",
  "AUTHOR" => "Damien \"dams\" Krotkine",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "MooX-LvalueAttribute",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooX::LvalueAttribute",
  "PREREQ_PM" => {
    "Hash::Util::FieldHash::Compat" => 0,
    "Moo" => "1.000008",
    "Variable::Magic" => "0.52",
    "strictures" => 1
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Devel::Hide" => 0,
    "Test::More" => "0.88",
    "blib" => 0
  },
  "VERSION" => "0.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



